### ISSUE TITLE
Websockets : 1006 Connection dropped by remote peer on chrome

### ISSUE THREAD

**Issue by anupamkumar** [theturtle32/WebSocket-Node#182]
Client Code:
```
<script>
            window.addEventListener("load",function(evt) {
                var chatbox = document.getElementById("chatbox");
                var send = document.getElementById("send");
                var chathistory = document.getElementById("chathistory");
                var socket = new WebSocket("ws://192.168.23.128:8000/wp",null);
                socket.addEventListener("open",function(evt) {
                    alert(JSON.stringify(evt));
                });
                socket.addEventListener("message",function(evt) {
                    chathistory.innerHTML += evt.data;
                    /*var msgType = jsonMsg['msgType'];
                    switch(msgType) {
                        case 'connectionstatus':
                            updatePlayerConnectionStatus(jsonMsg['payload']);
                            break;
                        case 'chatmsg':
                            updateChat(jsonMsg['payload']);
                            break;
                        case 'scoreupdate':
                            updateScores(jsonMsg['payload']);
                            break;
                    }*/     
                });

                send.addEventListener("click",function(evt) {
                    socket.send(chatbox.value);
                    chatbox.value = "";
                });
            });
        </script>
```

Server:

```
var server = http.createServer(function(req,res) {
...
...
}
server.listen(8000);
sessionmgr.sessiongc();
var cons= []
wsServer = new websockserver({
    httpServer: server,
    autoAcceptConnections: false
});
wsServer.on('request', function (req) {
    //console.log(req);
    //accept directly because authenticity of request's origin already established
    var con = req.accept(null, req.origin);
    cons.push(con);
    console.log(cons.length);
    console.log('connection accepted req origin = '+req.origin);
    con.on('message', function(msg) {
        if(msg.type === 'utf8') {
            console.log('Message:='+msg.utf8Data);
            con.sendUTF('Hello'+" "+msg);
        }
        else if(msg.type === 'binary') {
            console.log('Message:='+msg.binary);
            con.sendBytes(msg.binaryData);
        }
    });

    con.on('close',function(reasonCode, description) {
        console.log(con.remoteAddress +'disconnected :: '+reasonCode+" "+description);
    });
});
```

Firefox Works fine. Chrome errors out. 

Chrome Error Message:

```
WebSocket connection to 'ws://192.168.23.128:8000/wp' failed: Error during WebSocket handshake: Sent non-empty 'Sec-WebSocket-Protocol' header but no response was received
```

Chrome apparently handshakes differently, I guess, than Firefox and Node WebSocket Server is not handshaking with Chrome the way chrome wants ?

Any ideas on fixing for this ?

**Comment by anupamkumar** [comment]
anyone ? really wished my app would work on chrome ...

**Comment by theturtle32** [comment]
Try using something other than null for your subprotocol.

**Comment by braytak** [comment]
@anupamkumar what exactly are you trying to achieve with a null valued protocol field? Are you saying I am NOT negotiating for a subprotocol or are you saying I am negotiating for the null subprotocol?
In the former case, leave the field out. In the latter case, choose something other than null like @theturtle32  suggests.

FYI both Chrome and Firefox generate a "Sec-WebSocket-Protocol:null" request header. However the node server leaves out that header in its response. Chrome interprets this as meaning that the server does not agree to serve with the specified protocol while Firefox does not seem to care.

The standard is not totally clear as to the server behavior when the Sec-WebSocket-Protocol request header is null. The part that is however is that if Sec-WebSocket-Protocol is present in a request and the server does not agree with any of the *string* values of the header then it does not include that header in its response. So to me Chrome behaves correctly and both Chrome and Firefox should not include a Sec-WebSocket-Protocol in the request if the specified value is null.

In summary, stay away from a null valued Sec-WebSocket-Protocol.

**Comment by anupamkumar** [comment]
@theturtle32 , @braytak thanks for clearing this up. I asked this at stackoverflow too, but no one was able to explain it. 

It makes sense now.
